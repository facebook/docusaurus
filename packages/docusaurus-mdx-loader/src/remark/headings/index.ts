/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

/* Based on remark-slug (https://github.com/remarkjs/remark-slug) and gatsby-remark-autolink-headers (https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-remark-autolink-headers) */

import {parseMarkdownHeadingId, createSlugger} from '@docusaurus/utils';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer} from 'unified';
import type {Heading, Text} from 'mdast';

export interface PluginOptions {
  anchorsMaintainCase: boolean;
}

export default function plugin({
  anchorsMaintainCase,
}: PluginOptions): Transformer {
  return async (root) => {
    const {toString} = await import('mdast-util-to-string');
    const {visit} = await import('unist-util-visit');

    const slugs = createSlugger();
    visit(root, 'heading', (headingNode: Heading) => {
      const data = headingNode.data ?? (headingNode.data = {});
      const properties = (data.hProperties || (data.hProperties = {})) as {
        id: string;
      };
      let {id} = properties;

      if (id) {
        id = slugs.slug(id, {maintainCase: true});
      } else {
        const headingTextNodes = headingNode.children.filter(
          ({type}) => !['html', 'jsx'].includes(type),
        );
        const heading = toString(
          headingTextNodes.length > 0 ? headingTextNodes : headingNode,
        );

        // Support explicit heading IDs
        const parsedHeading = parseMarkdownHeadingId(heading);

        id =
          parsedHeading.id ??
          slugs.slug(heading, {maintainCase: anchorsMaintainCase});

        if (parsedHeading.id) {
          // When there's an id, it is always in the last child node
          // Sometimes heading is in multiple "parts" (** syntax creates a child
          // node):
          // ## part1 *part2* part3 {#id}
          const lastNode = headingNode.children[
            headingNode.children.length - 1
          ] as Text;

          if (headingNode.children.length > 1) {
            const lastNodeText = parseMarkdownHeadingId(lastNode.value).text;
            // When last part contains test+id, remove the id
            if (lastNodeText) {
              lastNode.value = lastNodeText;
            }
            // When last part contains only the id: completely remove that node
            else {
              headingNode.children.pop();
            }
          } else {
            lastNode.value = parsedHeading.text;
          }
        }
      }

      data.id = id;
      properties.id = id;
    });
  };
}
