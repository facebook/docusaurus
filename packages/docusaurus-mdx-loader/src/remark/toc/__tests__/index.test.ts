/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import vfile from 'to-vfile';
import plugin from '../index';
import headings from '../../headings/index';

const processFixture = async (name: string) => {
  const {default: gfm} = await import('remark-gfm');

  const {compile} = await import('@mdx-js/mdx');

  const filePath = path.join(
    __dirname,
    '__fixtures__',
    name.endsWith('.mdx') ? name : `${name}.md`,
  );

  const file = await vfile.read(filePath);

  const result = await compile(file, {
    format: 'mdx',
    remarkPlugins: [[headings, {anchorsMaintainCase: false}], gfm, plugin],
    rehypePlugins: [],
  });

  return result.value;
};

describe('toc remark plugin', () => {
  it('outputs empty array for no TOC', async () => {
    const result = await processFixture('no-heading');
    expect(result).toMatchSnapshot();
  });

  // A very implicit API: we allow users to hand-write the toc variable. It will
  // get overwritten in most cases, but until we find a better way, better keep
  // supporting this
  it('does not overwrite TOC var if no TOC', async () => {
    const result = await processFixture('no-heading-with-toc-export');
    expect(result).toMatchSnapshot();
  });

  it('works on non text phrasing content', async () => {
    const result = await processFixture('non-text-content');
    expect(result).toMatchSnapshot();
  });

  it('escapes inline code', async () => {
    const result = await processFixture('inline-code');
    expect(result).toMatchSnapshot();
  });

  it('works on text content', async () => {
    const result = await processFixture('just-content');
    expect(result).toMatchSnapshot();
  });

  it('exports even with existing name', async () => {
    const result = await processFixture('name-exist');
    expect(result).toMatchSnapshot();
  });

  it('inserts below imports', async () => {
    const result = await processFixture('insert-below-imports');
    expect(result).toMatchSnapshot();
  });

  it('handles empty headings', async () => {
    const result = await processFixture('empty-headings');
    expect(result).toMatchSnapshot();
  });

  it('works with imported markdown', async () => {
    const result = await processFixture('partials/index.mdx');
    expect(result).toMatchSnapshot();
  });

  it('works with partials importing other partials', async () => {
    const result = await processFixture('partials/_partial2.mdx');
    expect(result).toMatchSnapshot();
  });

  it('works with partial imported after its usage', async () => {
    const result = await processFixture(
      'partials/partial-used-before-import.mdx',
    );
    expect(result).toMatchSnapshot();
  });
});
