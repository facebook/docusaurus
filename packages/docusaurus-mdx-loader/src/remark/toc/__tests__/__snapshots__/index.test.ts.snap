// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toc remark plugin does not overwrite TOC var if no TOC 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = 1;
function _createMdxContent(props) {
  const _components = {
    code: "code",
    p: "p",
    pre: "pre",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.p, {
      children: "foo"
    }), "/n", _jsx(_components.p, {
      children: _jsx(_components.code, {
        children: "bar"
      })
    }), "/n", _jsx(_components.pre, {
      children: _jsx(_components.code, {
        className: "language-js",
        children: "baz/n"
      })
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin escapes inline code 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = [{
  "value": "<code>&lt;Head /&gt;</code>",
  "id": "head-",
  "level": 2
}, {
  "value": "<code>&lt;Head&gt;Test&lt;/Head&gt;</code>",
  "id": "headtesthead",
  "level": 3
}, {
  "value": "<code>&lt;div /&gt;</code>",
  "id": "div-",
  "level": 2
}, {
  "value": "<code>&lt;div&gt; Test &lt;/div&gt;</code>",
  "id": "div-test-div",
  "level": 2
}, {
  "value": "<code>&lt;div&gt;&lt;i&gt;Test&lt;/i&gt;&lt;/div&gt;</code>",
  "id": "divitestidiv",
  "level": 2
}, {
  "value": "<code>&lt;div&gt;&lt;i&gt;Test&lt;/i&gt;&lt;/div&gt;</code>",
  "id": "divitestidiv-1",
  "level": 2
}];
function _createMdxContent(props) {
  const _components = {
    a: "a",
    code: "code",
    h2: "h2",
    h3: "h3",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h2, {
      id: "head-",
      children: _jsx(_components.code, {
        children: "<Head />"
      })
    }), "/n", _jsx(_components.h3, {
      id: "headtesthead",
      children: _jsx(_components.code, {
        children: "<Head>Test</Head>"
      })
    }), "/n", _jsx(_components.h2, {
      id: "div-",
      children: _jsx(_components.code, {
        children: "<div />"
      })
    }), "/n", _jsx(_components.h2, {
      id: "div-test-div",
      children: _jsx(_components.code, {
        children: "<div> Test </div>"
      })
    }), "/n", _jsx(_components.h2, {
      id: "divitestidiv",
      children: _jsx(_components.code, {
        children: "<div><i>Test</i></div>"
      })
    }), "/n", _jsx(_components.h2, {
      id: "divitestidiv-1",
      children: _jsx(_components.a, {
        href: "/some/link",
        children: _jsx(_components.code, {
          children: "<div><i>Test</i></div>"
        })
      })
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin exports even with existing name 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = ['replaceMe'];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h2, {
      id: "thanos",
      children: "Thanos"
    }), "/n", _jsx(_components.h2, {
      id: "tony-stark",
      children: "Tony Stark"
    }), "/n", _jsx(_components.h3, {
      id: "avengers",
      children: "Avengers"
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin handles empty headings 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = [];
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    h2: "h2",
    img: "img",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h1, {
      id: "ignore-this",
      children: "Ignore this"
    }), "/n", _jsx(_components.h2, {
      id: ""
    }), "/n", _jsx(_components.h2, {
      id: "-1",
      children: _jsx(_components.img, {
        src: "an-image.svg",
        alt: ""
      })
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin inserts below imports 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import something from 'something';
import somethingElse from 'something-else';
export const toc = [{
  "value": "Title",
  "id": "title",
  "level": 2
}, {
  "value": "Test",
  "id": "test",
  "level": 2
}, {
  "value": "Again",
  "id": "again",
  "level": 3
}];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    p: "p",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h2, {
      id: "title",
      children: "Title"
    }), "/n", _jsx(_components.h2, {
      id: "test",
      children: "Test"
    }), "/n", _jsx(_components.h3, {
      id: "again",
      children: "Again"
    }), "/n", _jsx(_components.p, {
      children: "Content."
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin outputs empty array for no TOC 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = [];
function _createMdxContent(props) {
  const _components = {
    code: "code",
    p: "p",
    pre: "pre",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.p, {
      children: "foo"
    }), "/n", _jsx(_components.p, {
      children: _jsx(_components.code, {
        children: "bar"
      })
    }), "/n", _jsx(_components.pre, {
      children: _jsx(_components.code, {
        className: "language-js",
        children: "baz/n"
      })
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin works on non text phrasing content 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const toc = [{
  "value": "<em>Emphasis</em>",
  "id": "emphasis",
  "level": 2
}, {
  "value": "<strong>Importance</strong>",
  "id": "importance",
  "level": 3
}, {
  "value": "<del>Strikethrough</del>",
  "id": "strikethrough",
  "level": 2
}, {
  "value": "<i>HTML</i>",
  "id": "html",
  "level": 2
}, {
  "value": "<code>inline.code()</code>",
  "id": "inlinecode",
  "level": 2
}, {
  "value": "some <span class=\\"some-class\\">styled</span> <strong>heading</strong> <span class=\\"myClassName &lt;&gt; weird char\\"></span> test",
  "id": "some-styled-heading--test",
  "level": 2
}];
function _createMdxContent(props) {
  const _components = {
    code: "code",
    del: "del",
    em: "em",
    h2: "h2",
    h3: "h3",
    strong: "strong",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h2, {
      id: "emphasis",
      children: _jsx(_components.em, {
        children: "Emphasis"
      })
    }), "/n", _jsx(_components.h3, {
      id: "importance",
      children: _jsx(_components.strong, {
        children: "Importance"
      })
    }), "/n", _jsx(_components.h2, {
      id: "strikethrough",
      children: _jsx(_components.del, {
        children: "Strikethrough"
      })
    }), "/n", _jsx(_components.h2, {
      id: "html",
      children: _jsx("i", {
        children: "HTML"
      })
    }), "/n", _jsx(_components.h2, {
      id: "inlinecode",
      children: _jsx(_components.code, {
        children: "inline.code()"
      })
    }), "/n", _jsxs(_components.h2, {
      id: "some-styled-heading--test",
      children: ["some ", _jsx("span", {
        className: "some-class",
        style: {
          border: "solid"
        },
        children: "styled"
      }), " ", _jsx("strong", {
        children: "heading"
      }), " ", _jsx("span", {
        class: "myClass",
        className: "myClassName <> weird char",
        "data-random-attr": "456"
      }), " test"]
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin works on text content 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
export const c = 1;
export const toc = [{
  "value": "Endi",
  "id": "endi",
  "level": 3
}, {
  "value": "Endi",
  "id": "endi-1",
  "level": 2
}, {
  "value": "Yangshun",
  "id": "yangshun",
  "level": 3
}, {
  "value": "I â™¥ unicode.",
  "id": "i--unicode",
  "level": 2
}];
function _createMdxContent(props) {
  const _components = {
    code: "code",
    h2: "h2",
    h3: "h3",
    p: "p",
    pre: "pre",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h3, {
      id: "endi",
      children: "Endi"
    }), "/n", _jsx(_components.pre, {
      children: _jsx(_components.code, {
        className: "language-md",
        children: "## This is ignored/n"
      })
    }), "/n", _jsx(_components.h2, {
      id: "endi-1",
      children: "Endi"
    }), "/n", _jsx(_components.p, {
      children: "Lorem ipsum"
    }), "/n", _jsx(_components.h3, {
      id: "yangshun",
      children: "Yangshun"
    }), "/n", _jsx(_components.p, {
      children: "Some content here"
    }), "/n", _jsx(_components.h2, {
      id: "i--unicode",
      children: "I â™¥ unicode."
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin works with imported markdown 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import Partial1, {toc as __tocPartial1} from './_partial1.md';
import SomeComponent from './SomeComponent';
import Partial2, {toc as __tocPartial2} from './_partial2.md';
import UnusedPartialImport from './_partial3.md';
import DoesNotExist, {toc as __tocDoesNotExist} from './_doesNotExist.md';
export const toc = [{
  "value": "Index section 1",
  "id": "index-section-1",
  "level": 2
}, ...__tocPartial1, {
  "value": "Index section 2",
  "id": "index-section-2",
  "level": 2
}, ...__tocPartial2, {
  "value": "Unused partials",
  "id": "unused-partials",
  "level": 2
}, {
  "value": "NonExisting Partials",
  "id": "nonexisting-partials",
  "level": 2
}, ...__tocDoesNotExist, {
  "value": "Duplicate partials",
  "id": "duplicate-partials",
  "level": 2
}, ...__tocPartial1, ...__tocPartial1];
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    h2: "h2",
    p: "p",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h1, {
      id: "index",
      children: "Index"
    }), "/n", _jsx(_components.p, {
      children: "Some text"
    }), "/n", "/n", _jsx(_components.h2, {
      id: "index-section-1",
      children: "Index section 1"
    }), "/n", _jsx(_components.p, {
      children: "Foo"
    }), "/n", _jsx(Partial1, {}), "/n", _jsx(_components.p, {
      children: "Some text"
    }), "/n", _jsx(SomeComponent, {}), "/n", _jsx(_components.h2, {
      id: "index-section-2",
      children: "Index section 2"
    }), "/n", _jsx(Partial2, {}), "/n", _jsx(_components.h2, {
      id: "unused-partials",
      children: "Unused partials"
    }), "/n", _jsx(_components.p, {
      children: "Unused partials (that are only imported but not rendered) shouldn't alter the TOC"
    }), "/n", "/n", _jsx(_components.h2, {
      id: "nonexisting-partials",
      children: "NonExisting Partials"
    }), "/n", _jsx(_components.p, {
      children: "Partials that do not exist should alter the TOC"
    }), "/n", _jsx(_components.p, {
      children: "It's not the responsibility of the Remark plugin to check for their existence"
    }), "/n", "/n", _jsx(DoesNotExist, {}), "/n", _jsx(_components.h2, {
      id: "duplicate-partials",
      children: "Duplicate partials"
    }), "/n", _jsx(_components.p, {
      children: "It's fine if we use partials at the end"
    }), "/n", _jsx(Partial1, {}), "/n", _jsx(_components.p, {
      children: "And we can use the partial multiple times!"
    }), "/n", _jsx(Partial1, {})]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin works with partial imported after its usage 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import Partial, {toc as __tocPartial} from './_partial.md';
export const toc = [...__tocPartial];
function _createMdxContent(props) {
  const _components = {
    h1: "h1",
    p: "p",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h1, {
      id: "partial-used-before-import",
      children: "Partial used before import"
    }), "/n", _jsx(_components.p, {
      children: "While it looks weird to import after usage, this remains valid MDX usage."
    }), "/n", _jsx(Partial, {})]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;

exports[`toc remark plugin works with partials importing other partials 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
import Partial2Nested, {toc as __tocPartial2Nested} from './partial2-nested.md';
export const toc = [{
  "value": "Partial 2",
  "id": "partial-2",
  "level": 2
}, {
  "value": "Partial 2 Sub Heading",
  "id": "partial-2-sub-heading",
  "level": 3
}, ...__tocPartial2Nested];
function _createMdxContent(props) {
  const _components = {
    h2: "h2",
    h3: "h3",
    p: "p",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsx(_components.h2, {
      id: "partial-2",
      children: "Partial 2"
    }), "/n", _jsx(_components.p, {
      children: "Partial 2"
    }), "/n", _jsx(_components.h3, {
      id: "partial-2-sub-heading",
      children: "Partial 2 Sub Heading"
    }), "/n", _jsx(_components.p, {
      children: "Content"
    }), "/n", "/n", _jsx(Partial2Nested, {})]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;
